redash/extensions.py

```diff
@@ -1,30 +1,72 @@
-import os
-from pkg_resources import iter_entry_points, resource_isdir, resource_listdir
+from types import ModuleType
+from importlib_resources import contents, is_resource, path
+from importlib_metadata import entry_points
 
+BUNDLE_DIRECTORY = 'bundle'
 
-def init_extensions(app):
-    """
-    Load the Redash extensions for the given Redash Flask app.
+# The global Redash extension registry
+extensions = {}
+
+
+def resource_isdir(module, resource):
+    """Whether a given resource is a directory in the given module
+
+    https://importlib-resources.readthedocs.io/en/latest/migration.html#pkg-resources-resource-isdir
     """
-    if not hasattr(app, 'redash_extensions'):
-        app.redash_extensions = {}
+    try:
+        return resource in contents(module) and not is_resource(module, resource)
+    except (ImportError, TypeError):
+        # module isn't a package, so can't have a subdirectory/-package
+        return False
+
+
+def entry_point_module(entry_point):
+    """Returns the dotted module path for the given entry point"""
+    return entry_point.pattern.match(entry_point.value).group('module')
+
+
+def module_bundle_files(module):
+    if not resource_isdir(module, BUNDLE_DIRECTORY):
+        return
+
+    with path(module, BUNDLE_DIRECTORY) as bundle_dir:
+        # Copy content of extension bundle into extensions directory
+        return list(bundle_dir.rglob("*"))
 
-    for entry_point in iter_entry_points('redash.extensions'):
+
+def init_extensions(app):
+    """Load the Redash extensions for the given Redash Flask app.
+
+    The extension entry point can return any type of value but
+    must take a Flask application object.
+
+    E.g.::
+
+        def extension(app):
+            app.logger.info("Loading the Foobar extenions")
+            Foobar(app)
+
+    """
+    for entry_point in entry_points().get('redash.extensions', []):
         app.logger.info('Loading Redash extension %s.', entry_point.name)
+        module = entry_point_module(entry_point)
+        # First of all, try to get a list of bundle files
+        extensions[entry_point.name]['resource_list'] = module_bundle_files(module)
```

Comment:
Why not have a different entry point for the files (`redash.bundle` or something) to avoid all this logic?

Reply:
Hmm, I don't think there was a good reason to do it like this other than being conservative with the API design back in the day. Now that we've settled on multiple entry points it's a good idea to separate it, true. Thanks!

Reply:
This is done.